---
title: "Genetic Association Study with LazyGas"
author: "Your Name"
date: "`r Sys.Date()`"
output: html_document
---

## Introduction
This vignette demonstrates the process of conducting a genetic association study using the `lazyGas` package, along with the `GBScleanR` package for data cleaning. We will go through the steps of loading input files, preprocessing data, performing association analysis, and visualizing results. 

## Prerequisites
Ensure you have the following packages installed and loaded:
```{r}
library(lazyGas)
library(GBScleanR)
library(ggplot2)
```

## Loading Input GDS File and Building a LazyGas Object
Load the input GDS file and build a `LazyGas` class object:
```{r}
gds_fn <- system.file("extdata", "sample.gds", package = "lazyGas")
lg <- buildLazyGas(gds_fn = gds_fn, load_filter = TRUE, overwrite = FALSE)
```

If you have genotype information that are not stored in a GDS file, you can build 
a LazyGas class object by supplying a named list including the necessary 
information to the `create_gds` argument of the `buildLazyGas()` function.
```{r}
temp_dir <- tempdir()

# Now we assume genotype and marker data were provided as a matrix or 
# data.frame that have been loaded from CSV files.
genotype <- getGenotype(object = lg, node = "raw")
dosage <- getGenotype(object = lg, node = "dosage")
haplotype <- getHaplotype(object = lg)
snp.chromosome <- getChromosome(object = lg)
snp.position <- getPosition(object = lg)
snp.allele <- getAllele(object = lg)
sample.id <- getSamID(object = lg)
snp.id <- getMarID(object = lg)

create_gds <-  list(genotype = genotype,
                    sample.id = sample.id,
                    snp.id = snp.id,
                    snp.rs.id = NULL,
                    snp.chromosome = snp.chromosome,
                    snp.position = snp.position,
                    snp.allele = snp.allele,
                    haplotype = haplotype,
                    dosage = dosage)
lg <- buildLazyGas(gds_fn = file.path(temp_dir, "sample.gds"), 
                   create_gds = create_gds)
```


## Loading and Preprocessing Phenotype Data
Load the phenotype data:
```{r}
pheno_fn <- system.file("extdata", "pheno.csv", package = "lazyGas")
pheno <- read.csv(file = pheno_fn)
```

Assign phenotype data to the LazyGas object:
```{r}
lg <- assignPheno(object = lg,
                  pheno = pheno,
                  rename = "Fruite weight")
```

## Visualizing Phenotype Data
Generate and save plots for phenotype data:
```{r}
pheno <- getPheno(object = lg)
for(i in seq_along(pheno$pheno_names)){
  p <- plotPheno(object = lg, pheno = i, xlab = pheno$pheno_names[i])
  print(p)
}
```

<!-- ## Confirming dosage Data -->
<!-- Check dosage and sample IDs: -->
<!-- ```{r} -->
<!-- ds <- getGenotype(object = lg, node = "dosage", valid = FALSE) -->
<!-- all_sample_id <- getSamID(object = lg, valid = FALSE) -->
<!-- valid_indices <- validSam(object = lg) -->
<!-- all_sample_id[!valid_indices] # IDs of invalid (parental) samples -->
<!-- ds[!valid_indices, 1] # Assigned integers for the invalid (parental) samples -->
<!-- ``` -->

## Conducting Association Analysis
Define a function to convert dosage data to a model matrix:
```{r}
makeDF_FUN <- function(g){
  add <- g
  dom <- as.numeric(g == 1)
  out <- data.frame(add = add, dom = dom)
  colnames(out) <- c("add", "dom")
  return(out)
}
```

Run the association scan:
```{r}
scanAssoc(object = lg,
          formula = "add + dom",
          makeDF_FUN = makeDF_FUN,
          geno_format = "dosage",
          kruskal = NULL)
```

## Visualizing Results
Generate Manhattan plots for association results:
```{r}
pheno <- getPheno(object = lg)
for(i in seq_along(pheno$pheno_names)){
  p <- plotManhattan(object = lg, pheno = i)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}
```

Call and visualize peak blocks:
```{r}
callPeakBlock(object = lg, signif = 0.05, threshold = 0.8)

for(i in seq_along(pheno$pheno_names)){
  p <- plotPeaks(object = lg, pheno = i, recalc = FALSE)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}
```

## Recalculating and Visualizing Peaks
Recalculate and visualize peaks:
```{r}
recalcAssoc(object = lg, signif = 0.05, threshold = 0.8)

for(i in seq_along(pheno$pheno_names)){
  p <- plotPeaks(object = lg, pheno = i, recalc = TRUE)
  p <- p + labs(title = pheno$pheno_names[i])
  print(p)
}
```

## Visualizing Haplotype and Phenotype Data
Generate plots for haplotype and phenotype data:
```{r}
pdf("/home/ftom/hdd3/kosami/ponkan_ehime47/qtl/pheno_haplo_plot.pdf")
for(i in seq_along(pheno$pheno_names)){
  out <- haploPlot(object = lg, pheno = i, recalc = FALSE)
  print(out)
}
dev.off()

pdf("/home/ftom/hdd3/kosami/ponkan_ehime47/qtl/pheno_haplo_plot_recalc.pdf")
for(i in seq_along(pheno$pheno_names)){
  out <- haploPlot(object = lg, pheno = i, recalc = TRUE)
  print(out)
}
dev.off()
```

## Extracting Data
Extract data for further analysis:
```{r}
pheno <- getPheno(object = lg)
scan_pheno_1 <- lazyData(object = lg, dataset = "scan", pheno = pheno$pheno_names[1])
peakcall_pheno_5 <- lazyData(object = lg, dataset = "peakcall", pheno = pheno$pheno_names[5])
recalc_pheno_2 <- lazyData(object = lg, dataset = "recalc", pheno = pheno$pheno_names[2])
```

## List up candidate genes
You can list up candidate genes responsisble for the detected associations as 
following. 
```{r}

```

